torch/csrc/autograd/variable.cpp

```diff
@@ -17,89 +17,24 @@
 #include <stdexcept>
 #include <string>
 #include <vector>
+#include <typeinfo>
 
 namespace torch {
 namespace autograd {
-Variable::Impl::Impl(at::Tensor data, std::unique_ptr<Variable::AutogradMeta> autograd_meta, bool requires_grad, Edge gradient_edge)
-    : TensorImpl(data.type_id(), data.dtype(), data.device()),
-      data_(std::move(data)) {
-  autograd_meta->grad_fn_ = std::move(gradient_edge.function);
-  autograd_meta->requires_grad_ = false;
-  autograd_meta->is_view_ = false;
-  autograd_meta->output_nr_ = gradient_edge.input_nr;
+Variable::AutogradMeta::AutogradMeta(at::TensorImpl* self_impl, bool requires_grad, Edge gradient_edge) {
+  if (!self_impl) {
+    throw std::runtime_error("Attempted to set requires_grad of an undefined Variable");
```

Comment:
The original code `!data_.defined()` is called on the `data_` Tensor, and a Tensor's `defined()` function just returns its underlying `impl_`: https://github.com/pytorch/pytorch/blob/002009b5a99f225d898600f13c55087e2ab39cc9/aten/src/ATen/core/Tensor.h#L96-L98
Since `self_impl` is equivalent to `data_.defined()` in the original check, the equivalent check here would be to just use `!`.

Reply:
or just move the check to the place to the actual API on Tensor/Variable -- this isn't actually providing any safety because I could just pass some bogus value as `self_impl` and everything would still work.